### PR TITLE
Configurable reward functions

### PR DESCRIPTION
**Pull request by salmanmohammadi**
WandB run: https://wandb.ai/salman-mohammadi/torchtune/runs/9rkpzxwl

image

^ In other experiments with TRL I find small models struggle to capture the strict reward function, so further tuning may be required. I mainly added it to showcase how the user can have a nested reward function.

This PR provides support for configurable reward functions by defining a clear contract for users to define reward functions. It also adds support for fine-grained logging and reward functions which may be composed of multiple interactions. There are simple examples provided in the codebase, but we could consider more complex reward functions - for example, a reward function which uses output from a code interpreter may wish to provide separate rewards for whether code successfully compiles, and whether the code passes test cases, but doesn't try and run code which doesn't compile against test cases. In this case the user can easily monitor the success of each of these interactions in a single reward function class.

Future work should address more complex reward function spinup/teardown, and allow for resource allocation to reward functions (e.g. users may wish to load a reward model onto a specific device).

I think we should only provide minimal examples of working reward functions in the codebase which are sufficient to fully exercise the flexibility of the reward function design, and let users configure their own reward functions - I don't think we should try to maintain many reward functions.

@felipemello1

### CODE REVIEW

**Comment by SalmanMohammadi** [comment]
@felipemello1

**Comment by felipemello1** [comment]
@Darktex  @pbontrager

File: recipes/configs/dev/qwen3B_full_grpo.yaml [hunk 1]
```diff
@@ -121,6 +121,16 @@ vllm:
   steps_before_sync: 1
   queue_maxsize: ${eval:'${vllm.num_workers} * ${steps_before_sync}'}
 
+reward_functions:
```
Comment (Darktex) [inline recipes/configs/dev/qwen3B_full_grpo.yaml:124]: Love it, looks super clear. I was wondering whether we would need a registry for these, but I'm convinced that `cfg.instantiate` and a path are good enough.

File: torchtune/dev/grpo/rewards.py [hunk 2]
```diff
@@ -5,17 +5,148 @@
...
+        prefix = f"{prefix}/{self.reward_base_name}" if prefix else self.reward_base_name
+        for reward_name, reward in self.rewards.items():
+            log_dict[f"{prefix}/{reward_name}"] = reward.mean().item()
+        log_dict[f"{prefix}"] = self.total_reward.mean().item()
+        log_dict[f"{prefix}/successes"] = self.successes.mean().item()
+        return log_dict
+
+
+class RewardBase(Protocol):
+    def __call__(
```
Comment (Darktex) [inline torchtune/dev/grpo/rewards.py:35]: Taking a device in the call is perfect for these functions, but I think we should extend this protocol to support reward models. For a RM, we'd want to assign it to a device upon construction because we'd have to move it there and instantiate it. Where this gets tricky though is that you may have to shard even a reward model, so that should take a mesh.

Let's land this for now, and we'll re-evaluate down the road

File: torchtune/dev/grpo/rewards.py [hunk 3]
```diff
@@ -5,17 +5,148 @@
...
 
 from torchtune.modules.transforms.tokenizers import ModelTokenizer
 
+@dataclass
+class RewardOutput:
+    reward_base_name: str
+    total_reward: torch.Tensor
+    successes: torch.Tensor
+    # optional
+    rewards: Optional[Dict[str, torch.Tensor]] = field(default_factory=dict)
```
Comment (Darktex) [inline torchtune/dev/grpo/rewards.py:22]: I find this one confusing: what's the idea behind this? I have seen it used down in the formatter, but it seems to me that this is simply a "composite" reward. Why not simply make atomic functions? Rewards are all getting summed up anyway, so there doesn't seem any extra clarity in going with this approach?

File: torchtune/dev/rl/workers/trainers/pytorch_actor.py [hunk 4]
```diff
@@ -739,10 +739,10 @@ def _prepare_trajectory(self, raw_trajectory):
         seq_lens = raw_trajectory.seq_lens
         answers = raw_trajectory.answers
         policy_version = raw_trajectory.policy_version
-        rewards = raw_trajectory.rewards
+        # rewards = raw_trajectory.rewards
```
Comment (Darktex) [inline torchtune/dev/rl/workers/trainers/pytorch_actor.py:742]: This one we should keep: it's just the total reward

File: torchtune/dev/rl/workers/trainers/pytorch_actor.py [hunk 5]
```diff
@@ -739,10 +739,10 @@ def _prepare_trajectory(self, raw_trajectory):
...
         advantages = raw_trajectory.advantages
-        successes = raw_trajectory.successes
-        reward_metadata = raw_trajectory.reward_metadata
+        # successes = raw_trajectory.successes
```
Comment (Darktex) [inline torchtune/dev/rl/workers/trainers/pytorch_actor.py:744]: Similarly, this is the tally of successes

File: torchtune/dev/rl/workers/trainers/pytorch_actor.py [hunk 6]
```diff
@@ -739,10 +739,10 @@ def _prepare_trajectory(self, raw_trajectory):
...
+        # reward_metadata = raw_trajectory.reward_metadata
```
Comment (Darktex) [inline torchtune/dev/rl/workers/trainers/pytorch_actor.py:745]: Hmmm, this one is no longer needed as we now log it. I wonder if having a pointer to it in the trajectory is beneficial though

File: torchtune/dev/rl/workers/trainers/pytorch_actor.py (hunk 4 above)
Comment (Darktex) [reply torchtune/dev/rl/workers/trainers/pytorch_actor.py:742]: If we don't do monte carlo advantage estimation, then these are going to be used

**Review by Darktex — CHANGES_REQUESTED** [review]
Awesome PR Salman!!! Just a couple of changes and we are ready to merge!

The big one is basically not having "sub-rewards" in that `rewards` dict, but also happy to have more of a discussion about it if you see what this can support

File: recipes/configs/dev/qwen3B_full_grpo.yaml (hunk 1 above)
Comment (pbontrager) [reply recipes/configs/dev/qwen3B_full_grpo.yaml:124]: The problem with registries is that it doesn't see custom user code unless it's on the import path, so it forces users to work in the code base instead of just build on top of the code base.

File: torchtune/dev/grpo/rewards.py [hunk 7]
```diff
@@ -5,17 +5,148 @@
...
+        log_dict = {}
+        prefix = f"{prefix}/{self.reward_base_name}" if prefix else self.reward_base_name
+        for reward_name, reward in self.rewards.items():
+            log_dict[f"{prefix}/{reward_name}"] = reward.mean().item()
+        log_dict[f"{prefix}"] = self.total_reward.mean().item()
+        log_dict[f"{prefix}/successes"] = self.successes.mean().item()
+        return log_dict
+
+
+class RewardBase(Protocol):
```
Comment (pbontrager) [inline torchtune/dev/grpo/rewards.py:34]: I think this should just be called "Reward" without Base in the name. We haven't established a convention yet, but I don't think abstract classes should all be named base.

File: torchtune/dev/grpo/rewards.py [hunk 8]
```diff
@@ -5,17 +5,148 @@
...
+        for completion, answer in zip(completions, answers):
+            match = re.search(answer_pattern, completion, re.S)
+            if match:
+                answer = match.group(1).strip()
+                reward = self.positive_reward if answer == answer else self.negative_reward
+                rewards.append(reward)
+            else:
+                rewards.append(self.negative_reward)
+            
+        rewards = torch.tensor(rewards)
```
Comment (pbontrager) [inline torchtune/dev/grpo/rewards.py:78]: Do we not want to put this tensor on the given device, or is that not what the device is for?

File: torchtune/dev/grpo/rewards.py (hunk 7 above)
Comment (pbontrager) [reply torchtune/dev/grpo/rewards.py:34]: I also think that this should be an abstract base class instead of a protocol. Here you're mostly using it as an abc (since you keep inheriting from it), and in general I think abc's make it easier for users who are writing custom code to trace the interface they need to implement.

File: torchtune/dev/rl/workers/ref_actor.py [hunk 9]
```diff
@@ -153,32 +151,18 @@ def _log_metrics(
             }
         )
 
+        rewards_mean = torch.tensor([reward_output.total_reward.mean() for reward_output in reward_outputs]).mean()
```
Comment (pbontrager) [inline torchtune/dev/rl/workers/ref_actor.py:154]: I know this already existed in the code, but this is a risky summary metric as it's only valuable if your rewards all have similar scale.

**Review by pbontrager — COMMENTED** [review]
I left a few comments around naming and abstract classes but otherwise I love this PR. Thanks for doing this!

File: torchtune/dev/grpo/rewards.py (hunk 7 above)
Comment (Darktex) [reply torchtune/dev/grpo/rewards.py:34]: Sigh, this is such a mess on the Python side of things to be honest as there are like 95 ways of doing the same thing :-/ 

Technically, Protocols are supposed to be the most Pythonic way of going about this, but I agree that it looks cleaner if you explicitly inherit rather than duck type

File: torchtune/dev/rl/workers/ref_actor.py (hunk 9 above)
Comment (Darktex) [reply torchtune/dev/rl/workers/ref_actor.py:154]: Yeah but it doesn't alter behavior we already have so let's not put the onus of improving this on this PR. We can always have a follow-up PR that will target this behavior specifically.

BTW this is still a useful metric as the magnitude of the rewards _does_ affect learning dynamics, so if you weren't diligent in having them on similar scales that's on you :P

File: torchtune/dev/grpo/rewards.py (hunk 8 above)
Comment (pbontrager) [reply torchtune/dev/grpo/rewards.py:78]: Yeah, I think for reward functions that use a model, we might want to let the orchestration system place them, and then they can init themselves given the resources they can see locally. Definitely good to leave for followup

File: torchtune/dev/grpo/rewards.py (hunk 7 above)
Comment (pbontrager) [reply torchtune/dev/grpo/rewards.py:34]: I agree that it's very unclear, and so far, torchtune has done a poor job of enforcing interfaces. Even if protocols are the future for python, I'm wary of using them since it seems their primary purpose is for type support, but what I think torchtune needs are explicit interfaces that are easy for users to implement. I think abc's are more self documenting since every example you look at in the codebase will inherit from the abc.

Comment (Darktex) [reply torchtune/dev/grpo/rewards.py:34]: Fair, I agree (FWIW we do use protocols instead of abcs for recipes :P )

**Review by vmoens — COMMENTED** [review]
Thanks for this PR @SalmanMohammadi 
I'm not entirely convinced by this API for the following reasons:
- we need to make sure we have a reward format that works both online (during inference) and offline (during post-proc scoring). 
- it should be as easy to assign dense rewards (ie to every single token) than sparse like we do here. Sparse rewards are just a special instance of dense rewards.
- Similarly, the KL divergence penalty is just another reward and shouldn't have a special status.

My mental model around rewards is that rewards are first and foremost part of the environment. In some cases we want to assign rewards after the fact but that should not be the only thing we can do. 
In other words, out of our environments we should expect that we have rewards that are tensors that are shaped roughly identically as our tokens that we can then augment with some custom scores based on the model outputs.
Any scoring pipeline should assume that we (may) already have a reward (which can also be 0) that can also be composite, and that we are adding new items to it offline.

For example: if you could have a tool that is a kind of a black box and outputs rewards along its trajectory - I think that in this case you should log the rewards as rewards in your data structure and pass them to your scorer in the appropriate format. 
Another example is that you may have a reward pipeline that penalizes certain prohibited tokens in which case you need the reward to be dense.

In TorchRL, composite rewards are like composite observations or actions, they're just defined by a composite spec within the env and any transform / env that assigns composite rewards will have corresponding composite specs (specs are used to mark the input/output types of a given environment - they can be determined automatically).

One way you can work with composite rewards is:
```python
class RewardOutput(TensorClass["nocast"]):
    reward_base_name: str
    reward_val: torch.Tensor
    rewards: TensorDict = field(default_factory=TensorDict)

reward_spec = Composite(
    reward=Unbounded(...), # this is the "total" reward, summed over all sub-components
    reward_kl=Composite(
        reward_base_name=NonTensor(...), # name of the KL reward
        reward_val=Unbounded(...), # value of the component
        rewards=Composite(...),  # metadata / sub-components of the reward
        data_cls=RewardOutput
    ),
    ...
)
```

With this you can re-use the same API across your environment and your post-proc.

File: torchtune/dev/grpo/rewards.py [hunk 10]
```diff
@@ -5,17 +5,148 @@
...
+        return log_dict
+
+
+class RewardBase(Protocol):
+    def __call__(
+        self,
+        completion_ids: torch.Tensor,
+        completions: List[str],
+        answers: List[str],
+        device: torch.device,
```
[the remark this answers is not in the document]
Comment (SalmanMohammadi) [reply torchtune/dev/grpo/rewards.py:40]: Device is removed for now. The recipe hardcodes an answer so I think we should follow up and address further flexibility in the reward functions.

File: torchtune/dev/rl/workers/ref_actor.py (hunk 9 above)
Comment (SalmanMohammadi) [reply torchtune/dev/rl/workers/ref_actor.py:154]: relevant https://xcancel.com/vwxyzjn/status/1914779279918800968#m

File: pyproject.toml [hunk 11]
```diff
@@ -35,7 +35,6 @@ dependencies = [
 
     # Multimodal
     "Pillow>=9.4.0",
-    "math-verify[antlr4_9_3]"
```
Comment (SalmanMohammadi) [inline pyproject.toml:38]: I really don't think we should be taking this on as a dependency - users should feel free to configure their reward functions without requiring us to maintain them.

Comment (SalmanMohammadi) [reply pyproject.toml:38]: cc @joecummings @pbontrager

**Comment by SalmanMohammadi** [comment]
https://wandb.ai/salman-mohammadi/torchtune/runs/ngdpo2g2

Logging rewards in tables included

File: pyproject.toml (hunk 11 above)
Comment (Darktex) [reply pyproject.toml:38]: That package is how everyone scores their math answers though... Initially used by OpenAI, and adopted by everyone else. I think you do need it for a "reasonable" implementation of reasoning with math

Comment (salmanmohammadi) [reply pyproject.toml:38]: I can take it out of the core dependencies and into the async-rl for now, but our rewards should be examples rather than complete - I want pretty much all users to be defining their own reward functions.

**Comment by salmanmohammadi** [comment]
I'll move this PR over to torchtune once Davide's PR lands - thanks so much for the reviews.
